db/db_impl_write.cc

```diff
@@ -563,33 +559,41 @@ Status DBImpl::WriteImplWALOnly(const WriteOptions& write_options,
     }
     // else seq advances only by memtable writes
   }
-  if (status.ok() && write_options.sync) {
+  if (w.status.ok() && write_options.sync) {
     // Requesting sync with two_write_queues_ is expected to be very rare. We
     // hance provide a simple implementation that is not necessarily efficient.
     if (manual_wal_flush_) {
-      status = FlushWAL(true);
+      w.status = FlushWAL(true);
     } else {
-      status = SyncWAL();
+      w.status = SyncWAL();
     }
   }
   PERF_TIMER_START(write_pre_and_post_process_time);
 
   if (!w.CallbackFailed()) {
-    WriteCallbackStatusCheck(status);
-  }
-  nonmem_write_thread_.ExitAsBatchGroupLeader(write_group, w.status);
-  if (status.ok()) {
-    status = w.FinalStatus();
+    SetBackgroundErrorIfNeeded(w.status);
   }
+  // Propagate error, if any, to the whole write group
+  write_group.UpdateStatus(w.status);
+  nonmem_write_thread_.ExitAsBatchGroupLeader(write_group);
   if (seq_used != nullptr) {
     *seq_used = w.sequence;
   }
-  return status;
+  return w.FinalStatus();
```

Comment:
Previously w.FinalStatus would be returned if the status (w.status now) was Ok otherwise status (w.status) would be returned. But now we always return w.FinalStatus regardless. Can you explain why this change is ok?

Reply:
With the patch `w.status` is the same as previous `status`, which contain preprocess error, WAL error and memtable error. `w.FinalStatus` will return only be different from `w.status` when `w.status.ok() && !callback_status.ok()`. So the change here is  equivalent to previous version.